dissolve-struct-lib/src/main/scala/ch/ethz/dalab/dissolve/classification/MultiClassSVMWithDBCFW.scala

```diff
@@ -9,11 +9,41 @@ import breeze.linalg._
 import ch.ethz.dalab.dissolve.optimization.SolverUtils
 import ch.ethz.dalab.dissolve.optimization.DissolveFunctions
 import ch.ethz.dalab.dissolve.optimization.DBCFWSolverTuned
+import scala.collection.mutable.HashMap
+import org.apache.spark.rdd.PairRDDFunctions
 
 case class MultiClassLabel(label: Double, numClasses: Int)
 
 object MultiClassSVMWithDBCFW extends DissolveFunctions[Vector[Double], MultiClassLabel] {
 
+  val labelToWeight = HashMap[MultiClassLabel, Double]()
+
+  override def classWeights(label: MultiClassLabel): Double = {
+    labelToWeight.get(label).getOrElse(1.0)
+  }
+
+  def generateClassWeights(data: RDD[LabeledPoint]): Unit = {
+    val labels: Array[Double] = data.map { x => x.label }.distinct().collect()
+
+    val classOccur: PairRDDFunctions[Double, Double] = data.map(x => (x.label, 1.0))
+    val labelOccur: PairRDDFunctions[Double, Double] = classOccur.reduceByKey((x, y) => x + y)
+    val labelWeight: PairRDDFunctions[Double, Double] = labelOccur.mapValues { x => 1 / x }
+
+    val weightSum: Double = labelWeight.values.sum()
+    val nClasses: Int = labels.length
+    val scaleValue: Double = nClasses / weightSum
+
+    var sum: Double = 0.0
+    for ((label, weight) <- labelWeight.collectAsMap()) {
+      val clWeight = scaleValue * weight
+      sum += clWeight
+      labelToWeight.put(MultiClassLabel(label, labels.length), clWeight)
+    }
+
+    //make sure the weights sum up to the number of classes k
+    assert(sum == nClasses)
+  }
+
```

Comment:
can you avoid the code duplication with above? maybe put it in helpers or somewhere?